pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+	tv.mu.seenMsgStorageAppendResp = make(map[raftpb.PeerID]bool)
+	r.m.Store(te.Index, &tv)
+	return &tv
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te, nil, nil)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// MaybeRegister is called on an entry that is about to be proposed. This will
+// begin logging all subsequent updates to this entry. It will return true if
+// the registration succeeded.
+func (r *RaftTracer) MaybeRegister(ctx context.Context, ent raftpb.Entry) bool {
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEventf(ctx, 2, "skip registering raft proposal without index: %v", ent)
+		return false
+	}
+
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(ctx)
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntry to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(ctx, 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	ctx, span = r.tracer.StartSpanCtx(ctx, "raft trace",
+		tracing.WithParent(span), tracing.WithFollowsFrom())
+
+	tv := r.storeEntry(
+		kvserverpb.TracedEntry{
+			Index:   kvpb.RaftIndex(ent.Index),
+			TraceID: span.TraceID(),
+			SpanID:  span.SpanID(),
+		},
+		ctx,
+		span,
+	)
+	tv.logf(1, "registering local trace %s", tv)
+	return true
+}
+
+// MaybeTrace logs the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on the store.
+	if r.numRegisteredReplica.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// removeEntry remove this entry and decrements the registered counters at the
+// replica and store level.
+func (r *RaftTracer) removeEntry(index kvpb.RaftIndex) {
+	tv, found := r.m.LoadAndDelete(index)
+	if !found {
+		return
+	}
+	tv.mu.Lock()
+	defer tv.mu.Unlock()
+
+	if tv.mu.propSpan != nil {
+		tv.mu.propSpan.Finish()
```

Comment:
Should we also nil the `propSpan` here?

Reply:
I don't think it was strictly necessary, but changed anyway to make it clear.